cmd_actuators_mux/src/cmd_actuator_mux_nodelet.cpp

```diff
@@ -0,0 +1,393 @@
+
+  // Give permit to publish to this source if it's the only active or is
+  // already allowed or has higher priority that the currently allowed
+  if ((cmd_actuator_subs.allowed == VACANT) || (cmd_actuator_subs.allowed == idx) ||
+      (cmd_actuator_subs[idx]->priority > cmd_actuator_subs[cmd_actuator_subs.allowed]->priority))
+  {
+    if (cmd_actuator_subs.allowed != idx)
+    {
+      cmd_actuator_subs.allowed = idx;
+
+      // Notify the world that a new cmd_actuator source took the control
+      std_msgs::StringPtr acv_msg(new std_msgs::String);
+      acv_msg->data = cmd_actuator_subs[idx]->name;
+      active_subscriber.publish(acv_msg);
+    }
+    return true;
+  }
+  return false;
+}
+
+void CmdActuatorMuxNodelet::finAnglesTimerCallback(const ros::TimerEvent& event, unsigned int idx)
+{
+  timerCallbackProcess(cmd_fin_angles_subs, idx, fin_angles_active_subscriber);
+}
+
+void CmdActuatorMuxNodelet::setRPMtimerCallback(const ros::TimerEvent& event, unsigned int idx)
+{
+  timerCallbackProcess(cmd_set_rpm_subs, idx, set_rpm_active_subscriber);
+}
+
+void CmdActuatorMuxNodelet::timerCallbackProcess(CmdActuatorSubscribers& cmd_actuator_subs,
+                                                 const unsigned int& idx,
+                                                 const ros::Publisher& active_subscriber)
+{
+  if (cmd_actuator_subs.allowed == idx ||
+      (idx == GLOBAL_TIMER && cmd_actuator_subs.allowed != VACANT))
+  {
+    if (idx == GLOBAL_TIMER)
+    {
+      // No messages timeout happened for ANYONE, so last active source got stuck without further
+      // messages; not a big problem, just dislodge it; but possibly reflect a problem in the
+      // controller
+      NODELET_WARN("Actuator : No messages from ANY input received");
+      NODELET_WARN("Actuator: %s dislodged due to general timeout",
+                   cmd_actuator_subs[cmd_actuator_subs.allowed]->name.c_str());
+    }
+
+    // No messages timeout happened to currently active source, so...
+    cmd_actuator_subs.allowed = VACANT;
+
+    // ...notify the world that nobody is publishing; its vacant
+    std_msgs::StringPtr acv_msg(new std_msgs::String);
+    acv_msg->data = "idle";
+    active_subscriber.publish(acv_msg);
+  }
+
+  if (idx != GLOBAL_TIMER) cmd_actuator_subs[idx]->active = false;
+}
+
+void CmdActuatorMuxNodelet::onInit()
+{
+  nh = this->getNodeHandle();
+  pnh = this->getPrivateNodeHandle();
+
+  /*********************
+  ** Dynamic Reconfigure
+  **********************/
+  dynamic_reconfigure_cb = boost::bind(&CmdActuatorMuxNodelet::reloadConfiguration, this, _1, _2);
+  dynamic_reconfigure_server =
+      new dynamic_reconfigure::Server<cmd_actuators_mux::reloadConfig>(pnh);
+  dynamic_reconfigure_server->setCallback(dynamic_reconfigure_cb);
+
+  fin_angles_active_subscriber = pnh.advertise<std_msgs::String>("fin_angle_active", 1, true);
+  set_rpm_active_subscriber = pnh.advertise<std_msgs::String>("set_rpm_active", 1, true);
```

Comment:
@patapufate hmm, this suggests both inputs are independent. They should not be. If the fin commands are coming from the jaus ROS bridge, the mux should not relay thruster commands coming from the autopilot anymore.

Reply:
fixed in 899156c